### PR TITLE
Fixes #400, #653, #685, #686, 

### LINKED ISSUE #400: Kafka -> HBase pipeline
We would like to see a use case where the source is kafka and a sliding window within the DAG calculates and updates an average and then writes this to hbase. 
Subtasks:
- [ ] #685 Pipeline: Fix runtime classpath and fat jar assembly to take advantage of jars on target node
- [ ] #686 Pipeline: Define all schema data using upickle
- [ ] #687 Pipeline: Create HBase tables, columns or assume they are precreated and return appropriate error info to AppMaster
- [ ] #688 Pipeline: Dynamically set zookeeper, kafka and hbase server locations in respective taskactor's prestart's
- [ ] #689 Pipeline: Possibly add CLI options to pipeline for topic, tables which would override pipeline.conf
- [ ] #690 Pipeline: Define DAG streaming behavior
- [ ] #714 Pipeline: extend KafkaConfig -> PipelineConfig
- [ ] #724 Pipeline: add data generator to pipeline example

### PR DESCRIPTION
**Pull request by kkasravi**
Fixes 
#400 Kafka -> HBase pipeline 
#653 Start services in Yarn Application Master and register it's URL and the trackingURL
#685 Pipeline: Fix runtime classpath and fat jar assembly to take advantage of jars on target node
#686 Pipeline: Define all schema using upickle._

[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/702)

### CODE REVIEW

**Comment by kkasravi** [comment]
Note - @clockfly @huafengw  I'm am testing the fixes but wanted to make the PR available

File: services/src/main/scala/org/apache/gearpump/cluster/main/Services.scala [hunk 1]
```diff
@@ -37,18 +38,44 @@ object Services extends App with ArgumentsParser {
...
-    Console.out.println("Master URL: " + masterList)
+    Option(parse(args)).map(options => {
+      Option(options.getString("master")).map(masterList => {
+        LOG.info("Master URL: " + masterList)
 
-    implicit val system = ActorSystem("services" , ClusterConfig.load.application)
-    val master = system.actorOf(MasterProxy.props(Util.parseHostList(masterList)), MASTER)
+        implicit val system = ActorSystem("services", ClusterConfig.load.application)
+        val master = system.actorOf(MasterProxy.props(Util.parseHostList(masterList)), MASTER)
+        val config = system.settings.config
```
Comment (clockfly) [inline services/src/main/scala/org/apache/gearpump/cluster/main/Services.scala]: What have we changed?

Looks like doing the same thing.

File: services/src/main/scala/org/apache/gearpump/services/AppMasterService.scala [hunk 2]
```diff
@@ -42,8 +42,8 @@ import scala.util.{Failure, Success, Try}
 
 trait AppMasterService extends HttpService  {
   import upickle._
-  def master:ActorRef
   implicit val system: ActorSystem
+  implicit def master:ActorRef
```
Comment (clockfly) [inline services/src/main/scala/org/apache/gearpump/services/AppMasterService.scala]: Where is the implicit used?

File: services/src/main/scala/org/apache/gearpump/services/RestServices.scala [hunk 3]
```diff
@@ -20,36 +20,37 @@ package org.apache.gearpump.services
...
-  implicit def executionContext = actorRefFactory.dispatcher
+  implicit def executionContext: ExecutionContextExecutor = actorRefFactory.dispatcher
 
   lazy val routes = appMastersRoute ~ appMasterRoute ~ workersRoute ~ workerRoute ~
     masterRoute ~  configQueryRoute ~ metricQueryRoute ~ webSocketRoute ~ staticRoute
 }
 
-class RestServicesActor(masters: ActorRef, sys:ActorSystem) extends Actor with RestServices {
+class RestServicesActor(masters: ActorRef, sys:ActorSystem, conf:Config) extends Actor with RestServices {
   def actorRefFactory = context
```
Comment (clockfly) [inline services/src/main/scala/org/apache/gearpump/services/RestServices.scala]: conf can be retrieved by system.settings.config

Do we really need to add an extra parameter?

File: project/Build.scala [hunk 4]
```diff
@@ -451,7 +458,7 @@ object Build extends sbt.Build {
           "org.apache.hadoop" % "hadoop-yarn-server-nodemanager" % clouderaVersion % "provided"
         )
       )
-  ) dependsOn(core % "test->test", core % "provided")
+  ) dependsOn(services % "test->test;compile->compile", core % "provided")
```
Comment (clockfly) [inline project/Build.scala]: why depending on services?

File: project/Build.scala [hunk 5]
```diff
@@ -21,46 +25,49 @@ object Build extends sbt.Build {
    */
   val travis_deploy = taskKey[Unit]("use this after sbt assembly packArchive, it will rename the package so that travis deploy can find the package.")
   
+  val scalaVersionMajor = "scala-2.11"
```
Comment (clockfly) [inline project/Build.scala]: Please use the original order.

File: experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/AmActor.scala [hunk 6]
```diff
@@ -86,12 +84,25 @@ class AmActor(appConfig: AppConfig, yarnConf: YarnConfiguration) extends Actor {
...
       if(needMoreMasterContainersState) {
-        masterContainersStarted += 1        
+        masterContainersStarted += 1
         LOG.info(s"Currently master containers started : $masterContainersStarted/${appConfig.getEnv(GEARPUMPMASTER_CONTAINERS).toInt}")
-        requestWorkerContainersIfNeeded
+        requestWorkerContainersIfNeeded()
       } else {
         workerContainersStarted += 1
         LOG.info(s"Currently worker containers started : $workerContainersStarted/${appConfig.getEnv(WORKER_CONTAINERS).toInt}")
+        workerContainersStarted == workerContainersRequested match {
```
Comment (clockfly) [inline experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/AmActor.scala]: To simplify the problem, can we also start master actor in YarnAM container?

File: experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+
+  val taskParallelism = parallelism
+
+  private val source: TimeReplayableSource = new KafkaSource(taskContext.appName, taskId, taskParallelism,
+    kafkaConfig, msgDecoder)
+  private var startTime: TimeStamp = 0L
+
+  def testParser(): Unit = {
+    try {
+      var jsonData = "{\"source_id\":\"src-4\",\"event_ts\":\"2015-03-26T18:26:57.112853223Z\",\"metrics\":[{\"dimension\":\"CPU\",\"metric\":\"total\",\"value\":29005},{\"dimension\":\"CPU\",\"metric\":\"user\",\"value\":271},{\"dimension\":\"CPU\",\"metric\":\"sys\",\"value\":424},{\"dimension\":\"LOAD\",\"metric\":\"min\",\"value\":0.04},{\"dimension\":\"MEM\",\"metric\":\"free\",\"value\":15625940992},{\"dimension\":\"MEM\",\"metric\":\"used\",\"value\":144596992}]}"
```
Comment (clockfly) [inline experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala]: Please use  """   """"  pair to wrap the string, no need to escape anymore.

File: experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+  object Converters {
+    def convert[T](data: String)(implicit handler: Handler[T]): Datum[T] = {
+      handler.handle(data)
+    }
+  }
+
+}
+
+object Base64 {
+  private[this] val zero = Array(0, 0).map(_.toByte)
```
Comment (clockfly) [inline experiments/pipeline/src/main/scala/org/apache/gearpump/experiments/pipeline/Processors.scala]: I think you can use guava to do the base64 encoding stuff.

http://stackoverflow.com/a/18241113

**Comment by kkasravi** [comment]
I will resubmit based on new branch to simplify things. Sean I'll incorporate any comments that still apply.

**Pull request closed without merging**